bigbio/hub/hub_repos/bioid/bioid.py

```diff
@@ -0,0 +1,423 @@
+
+        # dl_manager can accept any type of nested list/dict and will give back the same structure with the url replaced with the path to local files.
+
+        # TODO: KEEP if your dataset is PUBLIC; remove if not
+        urls = _URLS[_DATASETNAME]
+        data_dir = dl_manager.download_and_extract(urls)
+
+        # Not all datasets have predefined canonical train/val/test splits.
+        # If your dataset has no predefined splits, use datasets.Split.TRAIN for all of the data.
+
+        return [
+            datasets.SplitGenerator(
+                name=datasets.Split.TRAIN,
+                # Whatever you put in gen_kwargs will be passed to _generate_examples
+                gen_kwargs={
+                    # TODO
+                    "data_dir": data_dir,
+                    "split": "train",
+                },
+            ),
+        ]
+
+    def load_annotations(self, path: str) -> Dict[str, Dict]:
+        """
+        We load annotations from `annotations.csv`
+        becuase the one in the BioC xml files have offsets issues.
+        """
+
+        df = pd.read_csv(path, sep=",")
+
+        df.fillna(-1, inplace=True)
+
+        annotations: Dict[str, Dict] = {}
+
+        for record in df.to_dict("records"):
+
+            article_id = str(record["don_article"])
+
+            if article_id not in annotations:
+                annotations[article_id] = {}
+
+            figure = record["figure"]
+
+            if figure not in annotations:
+                annotations[article_id][figure] = []
+
+            annotations[article_id][figure].append(record)
+
+        return annotations
+
+    def load_data(self, data_dir: str) -> List[Dict]:
+        """
+        Compose text from BioC files with annotations from `annotations.csv`.
+        We load annotations from `annotations.csv`
+        becuase the one in the BioC xml files have offsets issues.
+        """
+
+        text_dir = os.path.join(data_dir, "BioIDtraining_2", "caption_bioc")
+        annotation_file = os.path.join(data_dir, "BioIDtraining_2", "annotations.csv")
+
+        annotations = self.load_annotations(path=annotation_file)
+
+        data = []
+
+        for file_name in os.listdir(text_dir):
+
+            if file_name.startswith(".") or not file_name.endswith(".xml"):
+                continue
+
+            collection = bioc.load(os.path.join(text_dir, file_name))
+
+            for document in collection.documents:
+
+                item = document.infons
+
+                assert (
+                    len(document.passages) == 1
+                ), "Document contains more than one passage (figure caption). This is not expected!"
+
+                passage = document.passages[0]
+
+                article_id = document.infons["pmc_id"]
+                figure = document.infons["sourcedata_figure_dir"]
+
+                try:
```

Comment:
Why not checking key existence via an if-statement?

Reply:
[EAFP](https://docs.python.org/3/glossary.html#term-EAFP) vs [LBYL](https://docs.python.org/3/glossary.html#term-LBYL).